### DISCUSSION TITLE
Thoughts on removed extras

### DISCUSSION THREAD

**Discussion by dusty-phillips** [LazyVim/LazyVim#3215] [General]
Wanted to open a discussion on the removed extras. @folke gave a good rationale here: https://github.com/LazyVim/LazyVim/commit/4f4911ff95bc35438a2b8dd2d058b15f105f2ff1#commitcomment-142138015 but I feel like Extras could do with a bit more organizing now.

I agree with @echasnovski that these extras are vital for me and for many modal editing power users, but I’m 100% behind slimming down core.

I wonder if it wouldn’t be worth having a “recommended extras” extra that bundles some of the more popular extras all in one go. 

I’m not sure how you would decide what goes in it, though. I’m picturing stuff that is “missing” or “augmenting” functionality in core LazyVim (like mini-surround, mini-ai, dial.nvim) as opposed to stuff that “replaces” functionality in core (e.g mini.indentscope, aerial)

I imagine the long-term goal of “extras” is to support one-key installs of all the most popular plugins, but that reintroduces a “what do I *actually* need” problem for new users. So a “recommended extras” feature would be kind of helpful.

An interesting alternative would be to restructure extras to suggest that some of them are mutually exclusive. So you could have a “which file manager do you prefer: neo-tree, mini.filles, or oil” or “for added context would you like to add navic or treesitter-context?” Or “trouble/symbols-outline/aerial”.

**Comment by dpetka2001** [comment]
I don't have a problem with having an Extra `recommended`, which would install some recommended Extras in one go, but I believe it would be best to still give the user the option to manually toggle on/off any of the recommended Extras they might not want, because in the end I believe that `recommended` Extras will be somewhat based on personal preferences either way and can't accommodate all users' preferences.

**Comment by folke** [comment]
What if I would simply add a new group between `enabled` and `disabled` with `recommended` extras?

That way, they get more visibility and the user can easily pick the ones they want.


![image](https://github.com/LazyVim/LazyVim/assets/292349/7faee91a-9721-48ba-909c-53045dfdd9b7)

**Reply by dpetka2001** [reply]
So, this would only be a UI change and the manual toggle on/off would still be separate for each Extra, correct? I think this is a good solution.

**Reply by folke** [reply]
Yep, new recommended section with recommended (and currently disabled) plugins

**Comment by folke** [comment]
Something like this?

![image](https://github.com/LazyVim/LazyVim/assets/292349/5a9a3865-ed98-4973-bd69-0d40c7eafcf1)

@dpetka2001 @dusty-phillips any other extras we should include?

**Reply by dusty-phillips** [reply]
I like this solution!

I like those recommendations. Some others I suggest in LazyVim for Ambitious Developers:

* One of navic or treesitter-context
* mini-hipatterns
* util.project

I'd also suggest nvim-spider, but I don't think there's an extra for it yet.

**Reply by dusty-phillips** [reply]
If you're adding sections anyway, it might be useful to split out "coding", "editor", and "lang" as separate sections.

**Reply by folke** [reply]
Good idea on splitting out at least langauge support as well!
I'm also adding `recommended` detection based on root patterns and filetype for the current buffer / directory.
That way, languages that are availkable as an extra would also be shown.

```lua
  recommended = function()
    return LazyVim.extras.wants({
      ft = {
        "javascript",
        "javascriptreact",
        "javascript.jsx",
        "typescript",
        "typescriptreact",
        "typescript.tsx",
      },
      root = { "tsconfig.json", "package.json", "jsconfig.json", ".git" },
    })
  end,
```

**Reply by dpetka2001** [reply]
I would also recommend troublev3, because I like it a lot better with its recent additions of lualine component and also symbols side panel and I don't have to add additional plugins for that. Of course I have no idea on when exactly you're going to make troublev3 the main branch, so if that's in the plans for the foreseeable future then maybe this is not needed if it's going to become a core plugin soonish :)

PS: I'd recommend against `util.project` as the plugin itself has not been updated for over a year and LazyVim itself enables `manual_mode` because it used to mess up the LSP `root_dir` and I have seen even recently questions from users about that. These points make it hard for me at least to put it under `recommended`.

PS2: `mini-hipatterns` and `treesitter-context`, as @dusty-phillips mentioned already, I believe are good candidates for `recommended`. However, I would suggest that Folke should maintain the strictness under which he started slimming down the core plugins, as there will likely be other recommendations from other users and not everything should be in that section (that includes my own recommendations as well).

**Reply by dusty-phillips** [reply]
Now I want to find a different nvim project plugin. :-D

**Reply by dpetka2001** [reply]
I use `telescope-project.nvim`, which is quite simple and does its job. It needs to be configured with what directories to watch (no automatic `.git` detection as far as I know), but it has default `cd_scope` to `tab` and doesn't tend to mess with `root_dir`. In combination with `scope.nvim` (to keep buffers under tab scope), my usual workflow is to open a new Neovim tab to open different project and the global `pwd` doesn't get messed up. It's not so feature-rich as other plugins, but it does what it's supposed to do and my own needs are pretty simple and haven't bothered searching for alternatives.

**Reply by folke** [reply]
![image](https://github.com/LazyVim/LazyVim/assets/292349/7e705380-8cbf-43d9-8099-3c894663d21e)

**Reply by folke** [reply]
Just pushed all the changes. Let me know what you think!
Every language extra will now automatically be marked as `recommended` based on the current directory and the opened buffer.

The screenshot above was in a rust project that also had a `README.md`

**Reply by dpetka2001** [reply]
Looks awesome!!! 😃 Just a minor question. How are the `recommended languages` populated? Do you have an open `markdown`/`rust` buffer and that's why they're recommended?

PS: never mind Folke, you got me covered with your previous comment. I was writing when you answered already.

**Reply by folke** [reply]
Haha, so yes indeed, both based on the current buffer and/or the current directory.

**Reply by dusty-phillips** [reply]
It looks great. I just installed all the recommended ones. Even the ones I didn't know I needed. ;-) I'll get the book updated ASAP.

**Reply by radvil** [reply]
This looks dope @folke. Nice addition...
